File: SMO/ScriptTable/CS/ScriptTable/ScriptTable.cs
/*============================================================================
  File:    ScriptTable.cs

  Summary: Implements a sample SMO table scripting utility in C#.

  Date:    January 25, 2005
------------------------------------------------------------------------------
  This file is part of the Microsoft SQL Server Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
============================================================================*/

#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using System.Collections.Specialized;
using System.Text;

// SMO namespaces
using Microsoft.SqlServer.Management.Common;
using Microsoft.SqlServer.Management.Smo;
using Microsoft.SqlServer;
using Microsoft.SqlServer.MessageBox;

#endregion

namespace Microsoft.Samples.SqlServer
{
    partial class ScriptTable : Form
    {
        // Use the Server object to connect to a specific server
        private Server SqlServerSelection;

        public ScriptTable()
        {
            InitializeComponent();
        }

        private void ScriptTable_Load(object sender, System.EventArgs e)
        {
            ServerConnection ServerConn;
            ServerConnect scForm;
            DialogResult dr;

            // Display the main window first
            this.Show();
            Application.DoEvents();

            ServerConn = new ServerConnection();
            scForm = new ServerConnect(ServerConn);
            dr = scForm.ShowDialog(this);
            if ((dr == DialogResult.OK) &&
                (ServerConn.SqlConnectionObject.State == ConnectionState.Open))
            {
                SqlServerSelection = new Server(ServerConn);
                if (SqlServerSelection != null)
                {
                    this.Text = Properties.Resources.AppTitle + SqlServerSelection.Name;

                    // Refresh database list
                    ShowDatabases(true);
                }
            }
            else
            {
                this.Close();
            }
        }

        private void ScriptTable_FormClosed(object sender, System.EventArgs e)
        {
            if (SqlServerSelection != null)
            {
                if (SqlServerSelection.ConnectionContext.SqlConnectionObject
                    .State == ConnectionState.Open)
                {
                    SqlServerSelection.ConnectionContext.Disconnect();
                }
            }
        }

        private void ScriptTableButton_Click(System.Object sender,
            System.EventArgs e)
        {
            Cursor csr = null;
            Database db;
            StringCollection strColl;
            Scripter scrptr;
            SqlSmoObject[] smoObjects;
            Table tbl;
            Int32 count;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Use the selected database
                db = (Database)DatabasesComboBox.SelectedItem;
                if (db.Name.Length == 0)
                {
                    ExceptionMessageBox emb = new ExceptionMessageBox();
                    emb.Text = Properties.Resources.NoDatabaseSelected;
                    emb.Show(this);

                    return;
                }

                // Create scripter object
                scrptr = new Scripter(SqlServerSelection);
                if (ScriptDropCheckBox.CheckState == CheckState.Checked)
                {
                    scrptr.Options.ScriptDrops = true;
                }
                else
                {
                    scrptr.Options.ScriptDrops = false;
                }

                scrptr.DiscoveryProgress +=
                    new ProgressReportEventHandler(
                    this.ScriptTable_DiscoveryProgressReport);
                scrptr.ScriptingProgress +=
                    new ProgressReportEventHandler(
                    this.ScriptTable_ScriptingProgressReport);
                if (TablesComboBox.SelectedIndex >= 0)
                {
                    // Get selected table
                    smoObjects = new SqlSmoObject[1];
                    tbl = (Table)TablesComboBox.SelectedItem;
                    if (tbl.IsSystemObject == false)
                    {
                        smoObjects[0] = tbl;
                    }

                    if (DependenciesCheckBox.CheckState == CheckState.Checked)
                    {
                        scrptr.Options.WithDependencies = true;
                    }
                    else
                    {
                        scrptr.Options.WithDependencies = false;
                    }

                    strColl = scrptr.Script(smoObjects);

                    // Clear control
                    ScriptTextBox.Clear();
                    count = 0;
                    foreach (String str in strColl)
                    {
                        count++;
                        sbrStatus.Text = string.Format(
                            System.Globalization.CultureInfo.InvariantCulture,
                            Properties.Resources.AppendingScript, count,
                            strColl.Count);
                        ScriptTextBox.AppendText(str);
                        ScriptTextBox.AppendText(Properties.Resources.Go);
                    }
                }
                else
                {
                    ExceptionMessageBox emb = new ExceptionMessageBox();
                    emb.Text = Properties.Resources.ChooseTable;
                    emb.Show(this);
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                // Clean up
                sbrStatus.Text = Properties.Resources.Done;

                // Restore the original cursor
                this.Cursor = csr;
            }
        }

        private void ShowDatabases(bool selectDefault)
        {
            // Show the current databases on the server
            try
            {
                // Clear control
                DatabasesComboBox.Items.Clear();

                // Limit the properties returned to just those that we use
                SqlServerSelection.SetDefaultInitFields(typeof(Database),
                    new String[] { "Name", "IsSystemObject", "IsAccessible" });

                // Add database objects to combobox; the default ToString 
                // will display the database name
                foreach (Database db in SqlServerSelection.Databases)
                {
                    if (db.IsSystemObject == false && db.IsAccessible == true)
                    {
                        DatabasesComboBox.Items.Add(db);
                    }
                }

                if ((selectDefault == true) &&
                    (DatabasesComboBox.Items.Count > 0))
                {
                    DatabasesComboBox.SelectedIndex = 0;
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
        }

        private void ShowTables()
        {
            Database db;
            Cursor csr = null;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Clear the tables list
                TablesComboBox.Items.Clear();

                // Limit the properties returned to just those that we use
                SqlServerSelection.SetDefaultInitFields(typeof(Table),
                    new String[] { "Name", "CreateDate", "IsSystemObject" });

                // Show the current tables for the selected database
                if (DatabasesComboBox.SelectedIndex >= 0)
                {
                    db = (Database)DatabasesComboBox.SelectedItem;

                    foreach (Table tbl in db.Tables)
                    {
                        if (tbl.IsSystemObject == false)
                        {
                            TablesComboBox.Items.Add(tbl);
                        }
                    }

                    // Select the first item in the list
                    if (TablesComboBox.Items.Count > 0)
                    {
                        TablesComboBox.SelectedIndex = 0;
                    }
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                this.Cursor = csr;  // Restore the original cursor
            }
        }

        private void Databases_SelectedIndexChangedComboBox(System.Object
            sender, System.EventArgs e)
        {
            ShowTables();
            if (DatabasesComboBox.SelectedIndex >= 0)
            {
                ScriptTableButton.Enabled = true;
            }
            else
            {
                ScriptTableButton.Enabled = false;
            }
        }

        private void ScriptTable_DiscoveryProgressReport(System.Object sender,
            ProgressReportEventArgs e)
        {
            sbrStatus.Text = string.Format(
                System.Globalization.CultureInfo.InvariantCulture,
                Properties.Resources.Discovering, e.TotalCount, e.Total);
            sbrStatus.Refresh();
        }

        private void ScriptTable_ScriptingProgressReport(System.Object sender,
            ProgressReportEventArgs e)
        {
            sbrStatus.Text = string.Format(
                System.Globalization.CultureInfo.InvariantCulture,
                Properties.Resources.Scripting,
                e.TotalCount, e.Total);
            sbrStatus.Refresh();
        }
    }
}

File: SMO/ScriptTable/VB/ScriptTable/ScriptTable.vb
'============================================================================
'  File:    ScriptTable.vb
'
'  Summary: Implements a sample SMO table scripting utility in VB.NET.
'
'  Date:    January 25, 2005
'------------------------------------------------------------------------------
'  This file is part of the Microsoft SQL Server Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
'
'  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'============================================================================

Public Class ScriptTable
    ' Use the Server object to connect to a specific server
    Private SqlServerSelection As Server

    Private Sub ScriptTableButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ScriptTableButton.Click
        Dim csr As Cursor = Nothing
        Dim sDatabaseName As String
        Dim strColl As Specialized.StringCollection
        Dim scrptr As Scripter
        'Dim tmpSmoObjects() As SqlSmoObject
        Dim smoObjects() As SqlSmoObject
        Dim tbl As Table
        Dim count As Int32

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor

            ' Use the selected database
            sDatabaseName = DatabasesComboBox.Text
            If sDatabaseName.Length = 0 Then
                Dim emb As New ExceptionMessageBox()
                emb.Text = My.Resources.NoDatabaseSelected
                emb.Show(Me)

                Return
            End If

            ' Create scripter object
            scrptr = New Scripter(SqlServerSelection)
            If ScriptDropCheckBox.CheckState = CheckState.Checked Then
                scrptr.Options.ScriptDrops = True
            Else
                scrptr.Options.ScriptDrops = False
            End If

            AddHandler scrptr.DiscoveryProgress, _
                AddressOf Me.ScriptTable_DiscoveryProgressReport
            AddHandler scrptr.ScriptingProgress, _
                AddressOf Me.ScriptTable_ScriptingProgressReport

            If TablesComboBox.SelectedIndex >= 0 Then
                ' Get selected table
                smoObjects = New SqlSmoObject(0) {}
                tbl = CType(TablesComboBox.SelectedItem, Table)
                If tbl.IsSystemObject = False Then
                    smoObjects(0) = tbl
                End If

                If (DependenciesCheckBox.CheckState = CheckState.Checked) Then
                    scrptr.Options.WithDependencies = True
                Else
                    scrptr.Options.WithDependencies = False
                End If

                strColl = scrptr.Script(smoObjects)

                ' Clear control
                ScriptTextBox.Clear()
                count = 0
                For Each str As String In strColl
                    count += 1
                    sbrStatus.Text = String.Format( _
                        System.Globalization.CultureInfo.InvariantCulture, _
                        My.Resources.AppendingScript, count, strColl.Count)

                    ScriptTextBox.AppendText(str)
                    ScriptTextBox.AppendText(My.Resources.Go)
                Next
            Else
                Dim emb As New ExceptionMessageBox()
                emb.Text = My.Resources.ChooseTable
                emb.Show(Me)
            End If

        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)

        Finally
            ' Clean up
            sbrStatus.Text = My.Resources.Done

            ' Restore the original cursor
            Me.Cursor = csr
        End Try
    End Sub
    Private Sub Databases_SelectedIndexChangedComboBox(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles DatabasesComboBox.SelectedIndexChanged
        ShowTables()

        If DatabasesComboBox.SelectedIndex >= 0 Then
            ScriptTableButton.Enabled = True
        Else
            ScriptTableButton.Enabled = False
        End If
    End Sub

    Private Sub ScriptTable1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Dim ServerConn As ServerConnection
        Dim scForm As ServerConnect
        Dim dr As DialogResult

        ' Display the main window first
        Me.Show()
        Application.DoEvents()

        ServerConn = New ServerConnection
        scForm = New ServerConnect(ServerConn)
        dr = scForm.ShowDialog(Me)
        If dr = Windows.Forms.DialogResult.OK AndAlso ServerConn.SqlConnectionObject.State _
            = ConnectionState.Open Then
            SqlServerSelection = New Server(ServerConn)
            If Not (SqlServerSelection Is Nothing) Then
                Me.Text = My.Resources.AppTitle & SqlServerSelection.Name

                ' Refresh database list
                ShowDatabases(True)
            End If
        Else
            Me.Close()
        End If
    End Sub

    Private Sub ScriptTable1_FormClosed(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed
        If Not (SqlServerSelection Is Nothing) Then
            If SqlServerSelection.ConnectionContext.SqlConnectionObject.State _
                = ConnectionState.Open Then
                SqlServerSelection.ConnectionContext.Disconnect()
            End If
        End If
    End Sub

    Private Sub ScriptTable_DiscoveryProgressReport(ByVal sender As System.Object, ByVal e As ProgressReportEventArgs)
        sbrStatus.Text = String.Format( _
            System.Globalization.CultureInfo.InvariantCulture, _
            My.Resources.Discovering, e.TotalCount, e.Total)
        sbrStatus.Refresh()
    End Sub

    Private Sub ScriptTable_ScriptingProgressReport(ByVal sender As System.Object, ByVal e As ProgressReportEventArgs)
        sbrStatus.Text = String.Format( _
            System.Globalization.CultureInfo.InvariantCulture, _
            My.Resources.Scripting, e.TotalCount, e.Total)
        sbrStatus.Refresh()
    End Sub

    Private Sub ShowDatabases(ByVal selectDefault As Boolean)
        ' Show the current databases on the server
        Try
            ' Clear control
            DatabasesComboBox.Items.Clear()

            ' Limit the properties returned to just those that we use
            SqlServerSelection.SetDefaultInitFields(GetType(Database), _
                New String() {"Name", "IsSystemObject", "IsAccessible"})

            ' Add database objects to combobox; the default ToString will display the database name
            For Each db As Database In SqlServerSelection.Databases
                If db.IsSystemObject = False AndAlso db.IsAccessible = True Then
                    DatabasesComboBox.Items.Add(db)
                End If
            Next

            If (selectDefault = True) AndAlso _
                (DatabasesComboBox.Items.Count > 0) Then
                DatabasesComboBox.SelectedIndex = 0
            End If
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        End Try
    End Sub

    Private Sub ShowTables()
        Dim db As Database
        Dim csr As Cursor = Nothing

        Try
            csr = Me.Cursor   ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor   ' Display the waiting cursor

            ' Clear the tables list
            TablesComboBox.Items.Clear()

            ' Limit the properties returned to just those that we use
            SqlServerSelection.SetDefaultInitFields(GetType(Table), _
                New String() {"Name", "CreateDate", "IsSystemObject"})

            ' Show the current tables for the selected database
            If (DatabasesComboBox.SelectedIndex >= 0) Then
                ' Get database object from combobox
                db = CType(DatabasesComboBox.SelectedItem, Database)

                For Each tbl As Table In db.Tables
                    If tbl.IsSystemObject = False Then
                        TablesComboBox.Items.Add(tbl)
                    End If
                Next

                ' Select the first item in the list
                If (TablesComboBox.Items.Count > 0) Then
                    TablesComboBox.SelectedIndex = 0
                End If
            End If

        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            Me.Cursor = csr  ' Restore the original cursor
        End Try
    End Sub
End Class
